src/transaction.rs

```diff
@@ -0,0 +1,254 @@
+use crate::{
+    ColumnFamily, CreateIter, DBRawIterator, DBVector, Error, OptimisticTransactionDB, ReadOptions,
+};
+use ffi;
+use libc::{c_char, size_t};
+use std::marker::PhantomData;
+
+pub struct Transaction<'a> {
+    pub(crate) inner: *mut ffi::rocksdb_transaction_t,
+    pub(crate) db: PhantomData<&'a OptimisticTransactionDB>,
+    snapshot: Option<*const ffi::rocksdb_snapshot_t>,
+}
+
+impl<'a> Transaction<'a> {
+    pub(crate) fn new(
+        inner: *mut ffi::rocksdb_transaction_t,
+        snapshot: Option<*const ffi::rocksdb_snapshot_t>,
+    ) -> Transaction<'a> {
+        Transaction {
+            inner,
+            db: PhantomData,
+            snapshot,
+        }
+    }
+
+    /// commits a transaction
+    pub fn commit(&self) -> Result<(), Error> {
+        unsafe {
+            ffi_try!(ffi::rocksdb_transaction_commit(self.inner,));
+        }
+        Ok(())
+    }
+
+    /// Delete a key inside a transaction
+    ///
+    /// ColumnFamilyHandle: default
+    pub fn delete(&self, key: &[u8]) -> Result<(), Error> {
+        unsafe {
+            ffi_try!(ffi::rocksdb_transaction_delete(
+                self.inner,
+                key.as_ptr() as *const c_char,
+                key.len() as size_t,
+            ));
+            Ok(())
+        }
+    }
+
+    /// Delete a key inside a transaction
+    pub fn delete_cf(&self, cf: ColumnFamily, key: &[u8]) -> Result<(), Error> {
+        unsafe {
+            ffi_try!(ffi::rocksdb_transaction_delete_cf(
+                self.inner,
+                cf.inner,
+                key.as_ptr() as *const c_char,
+                key.len() as size_t,
+            ));
+            Ok(())
+        }
+    }
+
+    /// Read a key inside a transaction
+    ///
+    /// ColumnFamilyHandle: default
+    pub fn get_opt(
+        &self,
+        key: &[u8],
+        mut readopts: ReadOptions,
+    ) -> Result<Option<DBVector>, Error> {
+        if readopts.inner.is_null() {
```

Comment:
I feel like this check needs to be moved into a `ReadOptions::new` function that returns a result.  Maybe not something we need to do in the PR, but it feels like the wrong place for this check.